tinymod/plugins/code_metrics.py

```diff
@@ -0,0 +1,225 @@
+import math
+import shelve
+import hashlib
+import tempfile
+import os
+from datetime import datetime, timedelta
+from pathlib import Path
+import tokenize
+import token
+import gzip
+from io import BytesIO
+import pygal
+from pygal.style import NeonStyle
+from PIL import Image
+import glob
+import numpy as np
+from hata import Client, Guild, Role
+from hata.ext.slash import InteractionResponse
+from hata.ext import asyncio
+from scarletio import Lock, get_event_loop
+import aiosqlite
+import pickle
+
+TinyMod: Client
+GUILD: Guild
+ADMIN_ROLE: Role
+
+REPO = "https://github.com/tinygrad/tinygrad.git"
+BRANCH = "master"
+FILE_FILTER = "tinygrad/**/*.py"
+STYLE = NeonStyle(font_family="sans-serif", title_font_size=24, legend_font_size=18, background="#151510", plot_background="#151510")
+WORKING_DIR = Path(tempfile.gettempdir()) / f"tinymod-{hashlib.sha256(REPO.encode()).hexdigest()}"
+DATABASE = "tinymod.db"
+LOOP = get_event_loop()
+UPDATE_LOCK = Lock(LOOP)
+
+async def setup(_):
+  async with aiosqlite.connect(DATABASE) as db:
+    await db.execute("CREATE TABLE IF NOT EXISTS metrics(hash, date, raw)")
+    await db.commit()
+
+async def load_metrics():
+  async with aiosqlite.connect(DATABASE) as db:
+    return [ pickle.loads(item[0]) async for item in await db.execute("SELECT raw from metrics") ]
+
+async def insert_metrics(metrics):
+  async with aiosqlite.connect(DATABASE) as db:
+    await db.executemany("INSERT INTO metrics VALUES(?, ?, ?)", [ (m["hash"], m["date"], pickle.dumps(m)) for m in metrics ])
```

Comment:
aren't the metrics just some floats? is it possible to not use pickle here?

Reply:
Yh json should work fine. Will change and test.

Reply:
yes, but its a nested structure. It has one dict for the total codebase and one for each scanned file. I can move it to 2 tables to make the db structure "nice", will do this on the weekend. Increases code complexity though.

Reply:
Maybe a hybrid is a good idea. files as json and the rest in a table. The files are only used when generating the tables, meaning you will only have to deserialize it then and only a single one. For the graph the top structure is enough. I assume you are concerned about performance and not storage.

Reply:
yea mainly about perf, storage is cheap